File: LeetCode/Easy/ValidParentheses.java
/* 

Solution to Problem "Valid Parentheses" on LeetCode
A. K. Nguyen
Boston University 2022

Runtime: 557 ms
Memory Usage: 158.4 MB
 
*/

import java.util.*;
class ValidParentheses {
    public boolean isValid(String s) {
        
        Stack<Character> STACK = new Stack<Character>();
            
        for (int i = 0; i < s.length(); i++)
        {
            if (s.charAt(i) == '{' || s.charAt(i) == '(' || s.charAt(i) == '[')
            {
                STACK.push(s.charAt(i));
                System.out.println(Arrays.toString(STACK.toArray()));
            }
            else
            {
                if (STACK.size() == 0)
                {
                    return false;
                }
                char current_char = STACK.pop();
                if (current_char == '{')
                {
                    if (s.charAt(i) != '}')
                    {
                        return false;
                    }
                }
                if (current_char == '[')
                {
                    if (s.charAt(i) != ']')
                    {
                        return false;
                    }
                }
                if (current_char == '(')
                {
                    if (s.charAt(i) != ')')
                    {
                        return false;
                    }
                }
            
            }
            
        }
        if (STACK.size() != 0)
        {
            return false;
        }
        else
        {
            return true;
        }

        
    }
}

File: LeetCode/Easy/validparentheses.py
# Solution to Problem "Valid Parentheses" on LeetCode
# A. K. Nguyen
# Boston University 2022

def isValid(self, s):
        """
        :type s: str
        :rtype: bool

        Runtime: 15ms
        Memory: 13.5 MB
        """
        stack = []
        
        for char in s:
            if char in ["(", "{", "["]:
                stack.append(char)
            else:
                if len(stack) == 0:
                    return False
                current_char = stack.pop()
                if current_char == "(":
                    if char != ")":
                        return False
                if current_char == "{":
                    if char != "}":
                        return False
                if current_char == "[":
                    if char != "]":
                        return False

        if len(stack) != 0:
            return False
        return True

print(isValid("{}()"))
print(isValid("{}[}()"))
print(isValid("{()}"))
